src/ripple/basics/Slice.h

```diff
@@ -34,34 +37,55 @@
 
 namespace ripple {
 
-/** An immutable linear range of bytes.
+/** A linear range of bytes.
 
     A fully constructed Slice is guaranteed to be in a valid state.
     A Slice is lightweight and copyable, it retains no ownership
     of the underlying memory.
 */
-class Slice
+template<bool Mutable>
+class SliceImpl
 {
 private:
-    std::uint8_t const* data_ = nullptr;
+    using TData = typename std::conditional<Mutable, std::uint8_t*, std::uint8_t const*>::type;
+    using TVoidData = typename std::conditional<Mutable, void*, void const*>::type;
+    TData data_ = nullptr;
     std::size_t size_ = 0;
 
 public:
     using const_iterator = std::uint8_t const*;
 
     /** Default constructed Slice has length 0. */
-    Slice() noexcept = default;
+    SliceImpl() noexcept = default;
 
-    Slice (Slice const&) noexcept = default;
-    Slice& operator= (Slice const&) noexcept = default;
+    SliceImpl (SliceImpl const&) noexcept = default;
+    SliceImpl& operator= (SliceImpl const&) noexcept = default;
 
     /** Create a slice pointing to existing memory. */
-    Slice (void const* data, std::size_t size) noexcept
-        : data_ (reinterpret_cast<std::uint8_t const*>(data))
+    SliceImpl (TVoidData data, std::size_t size) noexcept
+        : data_ (reinterpret_cast<TData>(data))
         , size_ (size)
     {
     }
 
+    /** Can convert from a mutable slice to a non-mutable slice */
+    template<class T,
+             class = std::enable_if_t<!Mutable && std::is_same<T, SliceImpl<true>>::value>>
+    SliceImpl (/*SliceImpl<true>*/T const& rhs) noexcept
+        :SliceImpl{rhs.data(), rhs.size()}
+    {
+    }
+
+    /** Can assign from a mutable slice to a non-mutable slice */
+    template<class T,
+             class = std::enable_if_t<!Mutable && std::is_same<T, SliceImpl<true>>::value>>
+    SliceImpl& operator= (/*SliceImpl<true>*/T const& rhs) noexcept
```

Comment:
Remove the commented-out type.

Reply:
Will do.

Reply:
addressed in e71cfc80b1583e5055c9f4ed5828b2e7c7deeeb4